src/3d/symbols/qgspoint3dbillboardmaterial.cpp

```diff
@@ -0,0 +1,146 @@
+
+  // Technique
+  Qt3DRender::QTechnique *technique = new Qt3DRender::QTechnique;
+  technique->addRenderPass( renderPass );
+  technique->addFilterKey( filterKey );
+  technique->graphicsApiFilter()->setApi( Qt3DRender::QGraphicsApiFilter::OpenGL );
+  technique->graphicsApiFilter()->setProfile( Qt3DRender::QGraphicsApiFilter::CoreProfile );
+  technique->graphicsApiFilter()->setMajorVersion( 3 );
+  technique->graphicsApiFilter()->setMinorVersion( 1 );
+
+  // Effect
+  Qt3DRender::QEffect *effect = new Qt3DRender::QEffect( this );
+  effect->addTechnique( technique );
+
+  setEffect( effect );
+}
+
+void QgsPoint3DBillboardMaterial::setSize( const QSizeF size )
+{
+  mSize->setValue( size );
+}
+
+QSizeF QgsPoint3DBillboardMaterial::size() const
+{
+  return mSize->value().value<QSizeF>();
+}
+
+void QgsPoint3DBillboardMaterial::setViewportSize( const QSizeF size )
+{
+  mViewportSize->setValue( size );
+}
+
+QSizeF QgsPoint3DBillboardMaterial::windowSize() const
+{
+  return mViewportSize->value().value<QSizeF>();
+}
+
+void QgsPoint3DBillboardMaterial::setTexture2DFromImage( QImage image, double size )
+{
+  // Create texture image
+  QgsRectangle randomExtent = QgsRectangle( rand(), rand(), rand(), rand() );
+  QgsTerrainTextureImage *billboardTextureImage = new QgsTerrainTextureImage( image, randomExtent, QStringLiteral( "billboard material." ) );
+
+  setTexture2DFromTextureImage( billboardTextureImage );
+  setSize( QSizeF( size + size, size + size ) );
+}
+
+void QgsPoint3DBillboardMaterial::useDefaultSymbol( const Qgs3DMapSettings &map, bool selected )
+{
+  // Default texture
+  std::unique_ptr< QgsMarkerSymbol> defaultSymbol( static_cast<QgsMarkerSymbol *>( QgsSymbol::defaultSymbol( QgsWkbTypes::PointGeometry ) ) );
+  setTexture2DFromSymbol( defaultSymbol.get(), map, selected );
+}
+
+void QgsPoint3DBillboardMaterial::setTexture2DFromSymbol( QgsMarkerSymbol *markerSymbol, const Qgs3DMapSettings &map, bool selected )
+{
+  QgsRenderContext context;
+  context.setSelectionColor( map.selectionColor() );
+  context.setScaleFactor( map.outputDpi() / 25.4 );
+  double pixelSize = context.convertToPainterUnits( markerSymbol->size(),  markerSymbol->sizeUnit() );
+
+  QPixmap symbolPixmap = QgsSymbolLayerUtils::symbolPreviewPixmap( markerSymbol, QSize( int( pixelSize ), int( pixelSize ) ), 2, &context, selected );
+  QImage symbolImage = symbolPixmap.toImage();
+  QImage flippedSymbolImage = symbolImage.mirrored();
+  setTexture2DFromImage( flippedSymbolImage, pixelSize );
+}
+
+void QgsPoint3DBillboardMaterial::setTexture2DFromTextureImage( Qt3DRender::QAbstractTextureImage *textureImage )
+{
+  // Texture2D (memory leak here?)
+  Qt3DRender::QTexture2D *texture2D = new Qt3DRender::QTexture2D( this );
+  texture2D->setGenerateMipMaps( false );
+  texture2D->setMagnificationFilter( Qt3DRender::QTexture2D::Linear );
+  texture2D->setMinificationFilter( Qt3DRender::QTexture2D::Linear );
+
+  texture2D->addTextureImage( textureImage );
```

Comment:
answer to Nyall's question is yes - it does take ownership if the image does not have a parent yet - it would be good to have a comment here that the image gets parented to texture2D so it is clear...